include/oneapi/dpl/internal/async_extension_defs.h

```diff
@@ -0,0 +1,280 @@
+/*
+ *  Copyright (c) 2020 Intel Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef _ONEDPL_ASYNC_EXTENSION_DEFS_H
+#define _ONEDPL_ASYNC_EXTENSION_DEFS_H
+
+#include <CL/sycl.hpp>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace __internal
+{
+
+using sycl::event;
+
+template <typename _T, typename _Tp = sycl_iterator<sycl::access::mode::read_write, _T, sycl::buffer_allocator>>
+struct __async_value
+{
+    virtual _T
+    data() = 0;
+    virtual ~__async_value() = default;
+    virtual _Tp
+    raw_data() const = 0;
+};
+
+template <typename _T>
+struct __async_init : public __async_value<_T>
+{
+    _T __data;
+    __async_init(_T _d) : __data(_d) {}
+    _T
+    data()
+    {
+        return __data;
+    }
+    virtual sycl_iterator<sycl::access::mode::read_write, _T, sycl::buffer_allocator>
+    raw_data() const
+    {
+        return oneapi::dpl::begin(sycl::buffer<_T>{&this->__data, 1});
```

Comment:
What is the lifetime for `sycl::buffer` variable here?

Reply:
Up to the user, but we could also bind it to the lifetime of its future. In fact, this is a dummy because the virtual function `raw_data` needs a covariant return type. I decided to go with a `sycl_iterator` instead of a buffer that could be easily passed to another algorithm. The specific object holds the result of a reduction on an empty sequence.